### ISSUE TITLE
Add animation event support to scripted entities

### ISSUE THREAD

**Issue by Xyxen** [Facepunch/garrysmod-requests#529]
Scripted weapons support hooking into [animation events](https://developer.valvesoftware.com/wiki/Animation_Events) via [`SWEP:FireAnimationEvent`](http://wiki.garrysmod.com/page/WEAPON/FireAnimationEvent), but scripted entities don't seem to have any support for this. It should be as easy as running `DispatchAnimEvents(this)` in the entity's think method, and calling back into Lua in `HandleAnimEvent`.

Because the event might not be fired on the exact frame it's specified on (e.g. if the server FPS drops low), it would be ideal to pass the cycle of the event (`event.cycle`) to Lua. This way, the animation can be temporarily rewound if knowing the exact position of any moving attachment points on that frame is required.

**Comment by mcd1992** [comment]
+1

As it is we have to manually check the sequence every frame until we get what we're looking for. This is inefficient and pointless since valve has events in the engine already.

**Comment by mcd1992** [comment]
@robotboy655 Any update on the possibility of this? The current Lua workarounds are relatively CPU expensive and prone to desync between server / client.

**Comment by robotboy655** [comment]
HandleAnimEvent stuff is serverside only ( AE_ enums on VDK ) and not all of them seem to work ( tested on my TF2 Sentry Gun SENT, only the ```AE_CL_ENABLE_BODYGROUP``` worked ).

FireAnimationEvent is clientside, basically the same as the SWEP one, but it doesn't have the ```cycle``` information.

**Comment by Xyxen** [comment]
I understand that `HandleAnimEvent` is serverside (called via [`DispatchAnimEvents`](https://github.com/ValveSoftware/source-sdk-2013/blob/55ed12f8d1eb6887d348be03aee5573d44177ffb/mp/src/game/server/baseanimating.cpp#L1045-L1134)), and `FireEvent` is clientside (called via [`DoAnimationEvents`](https://github.com/ValveSoftware/source-sdk-2013/blob/55ed12f8d1eb6887d348be03aee5573d44177ffb/mp/src/game/client/c_baseanimating.cpp#L3333-L3516)). Both are very similar, and frustratingly different. `DoAnimationEvents` must have been written at a different time, because it inlines a slightly tweaked `GetAnimationEvent` and makes itself more confusing than it needs to be--and on top of that, fails to pass `pevent[i].cycle` to `FireEvent`.

Many other base animation events work (I've had success with [`AE_CL_PLAYSOUND`](https://github.com/ValveSoftware/source-sdk-2013/blob/55ed12f8d1eb6887d348be03aee5573d44177ffb/mp/src/game/client/c_baseanimating.cpp#L3726-L3740), as long as it's passed a manifest sound name). [`AE_CL_ENABLE_BODYGROUP`](https://github.com/ValveSoftware/source-sdk-2013/blob/55ed12f8d1eb6887d348be03aee5573d44177ffb/mp/src/game/client/c_baseanimating.cpp#L3879-L3887) and many of the rest have no reason to fail, either.

What I'm asking for, though, is a way to receive these events in Lua for scripted entities (at least on the server, but it would certainly be nice to have client support via `FireEvent` as well). For example:
```lua
-- init.lua
local MY_EVENT = 1
function ENT:OnAnimEvent(event, data, cycle)
    if (event == MY_EVENT) then
        self:Ignite(tonumber(data))
    end
end
```
Which would allow you to specify in a model's QC:
```
$sequence match_strike "anims/match_strike.smd" fps 30 {
  { event 1 10 "30" }
}
```
As a result, on (roughly) frame 10 of the "match_strike" sequence, event ID 1 is fired with the data "30", and the entity lights itself on fire for 30 seconds.

In my case, being passed the cycle is important because I rely on the position of a moving attachment at that frame. Animation events are fired when they lie between the current cycle and the last cycle animation events were processed on. This means that the server could stutter, and the cycle an animation resides on could be skipped over by (in an extreme case) a second or two. The event would be fired just fine, but the animation would be too far along for the attachment to be in the right place. Given the cycle of the event, it's simple enough to run
```lua
local oldCycle = self:GetCycle()
-- Perhaps check if the cycle difference is negligible
self:SetCycle(cycle)
self:FrameAdvance(0) --HACK because InvalidateBoneCache is client only
-- Do things with attachment points here, then restore oldCycle
```
Passing the cycle on the client could be painful. It would likely either involve modifying `DoAnimationEvent` and then propagating the change everywhere `FireEvent` is used, or copy/pasting the method into the scripted entity anim base and making the changes there. It's probably less of an issue on the client, so maybe it doesn't need to be done until a use arises?

**Comment by robotboy655** [comment]
ENT:FireAnimationEvent ( Clientside ) and ENT:HandleAnimEvent ( Serverside ) will be added in Dev branch, please test. Works only on ```ENT.Type = "anim"```
